pkg/component/spec.go

```diff
@@ -151,7 +151,9 @@ type ServiceOperationsSpec struct {
 
 // ServiceOperationsCommandSpec is the specification for execution of binaries to perform the check, install, and uninstall.
 type ServiceOperationsCommandSpec struct {
-	Args    []string         `config:"args,omitempty" yaml:"args,omitempty"`
-	Env     []CommandEnvSpec `config:"env,omitempty" yaml:"env,omitempty"`
-	Timeout time.Duration    `config:"timeout,omitempty" yaml:"timeout,omitempty"`
+	Args                   []string         `config:"args,omitempty" yaml:"args,omitempty"`
+	Env                    []CommandEnvSpec `config:"env,omitempty" yaml:"env,omitempty"`
+	Timeout                time.Duration    `config:"timeout,omitempty" yaml:"timeout,omitempty"`
+	RetryMaxCount          uint64           `config:"retry_max_count,omitempty" yaml:"retry_max_count,omitempty"`
```

Comment:
In the case of endpoint security, wouldn't it make more sense to keep retrying indefinitely? We should always be trying to converge the state of running components to what it should be.

Reply:
++ to retrying indefinitely for service commands.  We could leave the `RetryMaxCount` setting here and service components could simply not specify it in their spec file to achieve infinite retries OR we could not even provide this setting.  Which one do you think is better?  I lean towards the latter as it reduces any opportunities for misconfiguration and also reduces the API surface; we can always add this setting back later if it becomes useful.

---

With retrying indefinitely, I think we need to do one of two other things:
* Either cap the interval between retries to X duration, as this interval will grow indefinitely (at an exponential rate), or
* Reset the interval between retries to the original one every X retries

Do you have a preference?  

---

Finally, I'm starting to wonder if this retry functionality for service commands is strictly or largely a property of the Agent itself (as opposed to the service component).  That is, does it even make sense to allow service components to define retry-related settings in their spec files?  

The only retry-related setting  that I could see service components wanting to control is perhaps `RetrySleepInitDuration` as that might depend on how long a service command might take to process under normal circumstances.

Reply:
> Which one do you think is better? I lean towards the latter as it reduces any opportunities for misconfiguration and also reduces the API surface;

Agreed, removing it is simpler with less chance of mistakes.

> Either cap the interval between retries to X duration, as this interval will grow indefinitely (at an exponential rate), or
> Reset the interval between retries to the original one every X retries

I would just cap the interval between retries. I think https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff configures the default `MaxInterval` to 15m which is probably reasonable if we can confirm that.

> The only retry-related setting that I could see service components wanting to control is perhaps RetrySleepInitDuration as that might depend on how long a service command might take to process under normal circumstances

Yeah we will want to allow configuring the minimum interval between retries, and also set it higher than the default. The main reason is to protect the host system from being overloaded. Lets image a worst case scenario where there is a bug in Elastic Defend, and it uses 90% of the host CPU and then fails installing after the 10 minute timeout elapses. If we retry too quickly we'll just be pointlessly wasting resources.

Reply:
> I would just cap the interval between retries. I think https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff configures the default `MaxInterval` to 15m which is probably reasonable if we can confirm that.
> 

The default `MaxInterval` used by https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff is `60s`.  

https://github.com/cenkalti/backoff/blob/a04a6fe64ffb0e3fd0816460529d300be5f252df/exponential.go#L89
https://github.com/cenkalti/backoff/blob/a04a6fe64ffb0e3fd0816460529d300be5f252df/exponential.go#L79

I'll up it to `15m` in our code.

Reply:
Still need to add a test case for execution succeeding after retrying once, but I've implemented the suggestions from this thread.